### PR TITLE
GemJoin Integration Tests

### PR DESCRIPTION
**Pull request by naszam**

### CODE REVIEW

File: src/join.t.sol [hunk 1]
```diff
@@ -0,0 +1,1237 @@
...
+        GemJoin repJoin = new GemJoin(address(vat), "REP", address(rep));
+        assertEq(repJoin.dec(), 18);
+        giveAuth(address(vat), address(repJoin));
+
+        rep.approve(address(repJoin), uint256(-1));
+        assertEq(rep.balanceOf(address(this)), 100 ether);
+        assertEq(rep.balanceOf(address(repJoin)), 0);
+        assertEq(vat.gem("REP", address(this)), 0);
+        repJoin.join(address(this), 10 ether);
+        assertEq(rep.balanceOf(address(repJoin)), 10 ether);
```
Comment (gbalabasquer) [inline src/join.t.sol]: Missing:
`assertEq(rep.balanceOf(address(this)), 90 ether);`

Comment (gbalabasquer) [reply src/join.t.sol]: Seems to happen everywhere.

File: src/join.t.sol [hunk 2]
```diff
@@ -0,0 +1,1237 @@
...
+    }
+
+    function testGemJoin_AAVE() public {
+        giveTokens(address(aave), 100 * WAD);
+        GemJoin aaveJoin = new GemJoin(address(vat), "AAVE", address(aave));
+        assertEq(aaveJoin.dec(), 18);
+        giveAuth(address(vat), address(aaveJoin));
+
+        aave.approve(address(aaveJoin), uint256(-1));
+        assertEq(aave.balanceOf(address(this)), 100 * 10**18);
```
Comment (gbalabasquer) [inline src/join.t.sol]: Let's change them to `ether`.

File: src/join.t.sol [hunk 3]
```diff
@@ -0,0 +1,1237 @@
...
+    }
+
+    function testGemJoin_MATIC() public {
+        giveTokens(address(matic), 100 * WAD);
+        GemJoin maticJoin = new GemJoin(address(vat), "MATIC", address(matic));
+        assertEq(maticJoin.dec(), 18);
+        giveAuth(address(vat), address(maticJoin));
+
+        matic.approve(address(maticJoin), uint256(-1));
+        assertEq(matic.balanceOf(address(this)), 100 * 10**18);
```
Comment (gbalabasquer) [inline src/join.t.sol]: Same in this one.

File: src/join.t.sol [hunk 4]
```diff
@@ -0,0 +1,1237 @@
...
+        assertEq(vat.gem("GUSD", address(this)), 6 ether);
+    }
+
+    function testGemJoin_PAXG() public {
+        giveTokens(address(paxg), 100 * WAD);
+        GemJoin9 paxgJoin = new GemJoin9(address(vat), "PAXG", address(paxg));
+        giveAuth(address(vat), address(paxgJoin));
+
+        paxg.approve(address(paxgJoin), uint256(-1));
+        assertEq(paxg.balanceOf(address(this)), 100 * 10 ** 18);
```
Comment (gbalabasquer) [inline src/join.t.sol]: Same here.

**Comment by gbalabasquer** [comment]
A part from the missing check I'd suggest standardizing the use of `10 ** 18` , `ether` and `WAD` to just one of them.

File: src/join.t.sol (hunk 1 above)
Comment (naszam) [reply src/join.t.sol]: I just did a [diffcheck](https://www.diffchecker.com/JAW3osv3) against `dss-gem-joins` `join.t.sol` and it seems the check was also missing there, as discussed in chat.

File: src/join.t.sol (hunk 2 above)
Comment (naszam) [reply src/join.t.sol]: Done in ef1fac7, using `WAD`

File: src/join.t.sol (hunk 3 above)
Comment (naszam) [reply src/join.t.sol]: Done.

File: src/join.t.sol (hunk 4 above)
Comment (naszam) [reply src/join.t.sol]: Done.

File: src/join.t.sol (hunk 1 above)
Comment (naszam) [reply src/join.t.sol]: Done in 883d8d2

File: src/join.t.sol [hunk 5]
```diff
@@ -285,6 +285,7 @@ contract GemJoinTest is DSTest {
         assertEq(vat.gem("REP", address(this)), 0);
         repJoin.join(address(this), 10 * WAD);
         assertEq(rep.balanceOf(address(repJoin)), 10 * WAD);
+        assertEq(rep.balanceOf(address(this)), 90 * WAD);
```
Comment (gbalabasquer) [inline src/join.t.sol:287]: Very minor thing, so feel free to change or not. The order of the checks is different before and after `exit`.

Comment (naszam) [reply src/join.t.sol:287]: Done in 1de6be7

**Review by gbalabasquer — APPROVED** [review]
LGTM

**Pull request merged**